### ISSUE TITLE
Sacrificing material over a check

### ISSUE THREAD

**Issue by ArcticXWolf** [Wikunia/Ghess#4]
Hey,

in a game our engines played a few minutes ago, your engine went for a check, but sacrificed a piece for that.

I'm sure this will be hard to track down (it might have something to do with the QSearch depth or else), but maybe the position will help you!

Game: https://lichess.org/ECQ1WIJu/white#109
Move 55, Black plays Bf2+, which leads to the capture Rxf2. Black cannot capture the other bishop, as that would lead to a backrank mate.

EDIT:
Btw, I'm wondering: what time control/calculation do you use?

![image](https://user-images.githubusercontent.com/5812215/115745297-975e6d00-a393-11eb-9438-843337c68a0c.png)

There is a bit of jitter in the move times, thus I'm not sure if you use a continous function to calculate the time ^^
Just curious

EDIT2: Nevermind, the peaks of the movetime chart show a continous function, the jitter comes from forced moves/simple decisions pruning a lot of time ^^

**Comment by Wikunia** [comment]
Thanks for opening the issue. I will have a deeper look later but it might be because of my quiescence search which only looks at captures and not checks. 

Regarding time management. I experienced with that a bit and basically break earlier if I assume I can't compute the next depth and in time and the drop are most likely because the engine thinks while yours makes a move and if that is saved to a depth were it would take too long to calculate deeper it just makes that move. In general the engine currently also overestimates the time it needs such that in general I think it should think a bit longer per move.  

Regarding edit2: yeah it's also true that I just make a move if it's forced.

**Comment by ArcticXWolf** [comment]
Ah, you think during the turn of the opponent? Nice!

How do you do it, do you use the ponder feature of uci or do you calculate "all" possible answers?

I guess the repo isnt up to date with your current lichess version, is it?

**Comment by Wikunia** [comment]
Yeah I have to push the latest stuff..a bit busy these days...
I calculate the principal variation and just take the most likely move of the opponent. I have to admit that I didn't really get what UCI is doing with this bestmove ... ponder ... stuff so I just do it internally.

**Comment by Wikunia** [comment]
So currently it only works if I get the UCI stuff for a single game in order

**Comment by ArcticXWolf** [comment]
Yeah, sadly the UCI specification is quite vague. My understanding is the following:
the gui (aka other party in the uci-protocol) can instruct the the engine to ponder during the opponents turn via `go ponder`. Then the engine is expected to compute from the pondermove (last move in move list of the position command <- this move has not been made by the opponent yet) as far as it can. As soon as the opponent makes their move, the GUI sends ponderhit (if the opponent actually makes the move to ponder on) or start a new search (if the opponent cchose a different one). On a ponderhit the engine shall just continue from the already computed tree doing a normal search.

And the ponder option in bestmove is just a way for the engine to signal which move it wants to ponder on (most likely via the pv) to the GUI. However the GUI is free to instruct a different ponder move if it wants to (strange decision, I think).

**Comment by Wikunia** [comment]
Oh I see. Thanks for the explanation. That makes more sense. I'll definitely change to that system in the future.
